packages/react-query/src/HydrationBoundary.tsx

```diff
@@ -19,19 +23,83 @@ export const HydrationBoundary = ({
   queryClient,
 }: HydrationBoundaryProps) => {
   const client = useQueryClient(queryClient)
+  const [hydrationQueue, setHydrationQueue] = React.useState<
+    DehydratedState['queries'] | undefined
+  >()
 
   const optionsRef = React.useRef(options)
   optionsRef.current = options
```

Comment:
nit: I would put this in a useEffect that always runs, as writing to refs during render is technically also not pure

```suggestion
  React.useEffect(() => {
    optionsRef.current = options
  })  
```

Reply:
Yeah, I've been staring at that ref trying to understand it. I'm not sure if I wrote it at the beginning or if it happened later, but if the purpose is to not retrigger hydration just because the options have changed, we can do that by just ignoring it in the dep arrays (rarely great, but better than a ref in that case).

With `useEffect` there's also this new case where another fiber could render this component with other options and that would translate over to the effect in "this" fiber (this is one unsafe part about refs I guess).

If we just use the prop as is, and add it to the dep arrays, that would require users to memoize it or it would cause rerunning hydration on every render.

Doing it in an effect like this means new options wont be applied to the hydrations that happen in render though, so not sure that's great either?

Reply:
> we can do that by just ignoring it in the dep arrays

wouldn't do that. if options contain functions that could closure over something from a previous render, you'll have a stale closure problem.

I got the pattern to write to often changing objects or functions into a ref and update them in an effect from Kent, see: https://epicreact.dev/the-latest-ref-pattern-in-react/

we do the same e.g. here:

https://github.com/TanStack/query/blob/1d3ee672d2a1536e3463243813c91902aa97d973/packages/react-query/src/useMutationState.ts#L60-L62

Reply:
Yeah, it's a nice pattern sometimes, but that case in `useMutationState` works quite differently than this does. Here, if we used that pattern, the `optionsRef` wouldn't update until _after_ render, but we need the updated value inside of render when we are hydrating new queries.

First render:

```tsx
<HydrationBoundary state={queryAState} options={{ defaultOptions: { queries: { gcTime: 10000 } } }} />
```

On server components navigation:

```tsx
<HydrationBoundary state={queryBState} options={{ defaultOptions: { queries: { gcTime: 1000 } } }} />
```

Because `queryBState` is not already in the cache, it will get hydrated inside of render with the old 10000 value, since options and `gcTime` has not had a time to change to the new value which would now happen in an effect.

Reply:
ah yes, because we hydrate in useMemo as well. As usual, you are right 👍